File: plugins/fanqie.lua
local common = require("_shared.common")

local plugin = {}

plugin.manifest = {
    id = "fanqie",
    name = "番茄小说",
    version = "1.1.0",
    author = "fanqie-downloader-tui",
    description = "默认番茄小说书源插件，会自行从环境变量或 .env 中读取 FANQIE_APIKEY",
    required_envs = {
        "FANQIE_APIKEY",
    },
    optional_envs = {},
}

local API_URL = "http://v3.rain.ink/fanqie/"

local ctx = {
    api_key = "",
}

local function parse_book(data)
    return {
        book_id = common.get_string(data, "book_id", ""),
        title = common.get_string(data, "book_name", common.get_string(data, "title", "")),
        author = common.get_string(data, "author", ""),
        cover_url = common.get_string(data, "thumb_url", common.get_string(data, "cover_url", "")),
        abstract = common.get_string(data, "abstract", ""),
        category = common.get_string(data, "category", ""),
        word_count = common.get_string(data, "word_number", common.get_string(data, "word_count", "")),
        score = common.get_number(data, "score", 0),
        gender = common.get_number(data, "gender", 0),
        creation_status = common.get_number(data, "creation_status", 0),
        last_chapter_title = common.get_string(data, "last_chapter_title", ""),
        last_update_time = common.get_number(data, "last_chapter_update_time", 0),
    }
end

local function ensure_api_key()
    if ctx.api_key == "" then
        ctx.api_key = common.require_env("FANQIE_APIKEY")
    end
    return ctx.api_key
end

local function build_url(type_id, params)
    local query = {
        apikey = ensure_api_key(),
        type = type_id,
    }

    if params ~= nil then
        for key, value in pairs(params) do
            query[key] = value
        end
    end

    return common.append_query(API_URL, query)
end

function plugin.configure()
    ctx.api_key = common.require_env("FANQIE_APIKEY")
end

function plugin.search(keywords, page)
    local root = common.get_json(build_url(1, {
        keywords = keywords,
        page = page * 10,
    }))

    local results = {}

    if root.search_tabs ~= nil then
        for _, tab in ipairs(root.search_tabs) do
            if tab.data ~= nil then
                for _, item in ipairs(tab.data) do
                    if item.book_data ~= nil and item.book_data[1] ~= nil then
                        table.insert(results, parse_book(item.book_data[1]))
                    end
                end
                if #results > 0 then
                    break
                end
            end
        end
    elseif root.data ~= nil then
        for _, item in ipairs(root.data) do
            if item.book_data ~= nil and item.book_data[1] ~= nil then
                table.insert(results, parse_book(item.book_data[1]))
            else
                table.insert(results, parse_book(item))
            end
        end
    end

    return results
end

function plugin.get_book_info(book_id)
    local root = common.get_json(build_url(2, {
        bookid = book_id,
    }))
    if root.data == nil then
        return nil
    end
    return parse_book(root.data)
end

function plugin.get_toc(book_id)
    local root = common.get_json(build_url(3, {
        bookid = book_id,
    }))

    local results = {}
    local items = root.data and root.data.item_data_list or nil
    if items == nil then
        return results
    end

    for _, item in ipairs(items) do
        table.insert(results, {
            item_id = common.get_string(item, "item_id", ""),
            title = common.get_string(item, "title", ""),
            volume_name = common.get_string(item, "volume_name", ""),
            word_count = common.get_number(item, "chapter_word_number", 0),
            update_time = common.get_number(item, "first_pass_time", 0),
        })
    end

    return results
end

function plugin.get_chapter(book_id, item_id)
    local root = common.get_json(build_url(4, {
        itemid = item_id,
    }))
    if root.data == nil or root.data.content == nil then
        return nil
    end
    return {
        item_id = item_id,
        title = "",
        content = tostring(root.data.content),
    }
end

return plugin

File: plugins/fanqie.js
const common = require("_shared/common");
const rainBatch = require("_shared/rain_batch");

const API_URL = "http://v3.rain.ink/fanqie/";
const WEB_BASE_URL = "https://v3.rain.ink";
const LOGIN_URL = `${WEB_BASE_URL}/web/index.php`;
const COVER_ORIGIN_URL = "https://p6-novel.byteimg.com/origin/";
const INVALID_KEY_MARKER = "无效的API密钥";
const USER_RE = /const user = (\{.*?\});/s;

const ctx = {
  api_key: "",
  logged_in: false,
  user_info: null,
};

const normalizeCoverUrl = (value) => String(value || "").trim();

const isDisplayableCoverUrl = (value) => {
  const url = normalizeCoverUrl(value);
  return !!url && !/\.hei[cf](?:$|[?#])/i.test(url);
};

const buildOriginCoverUrl = (thumbUri) => {
  const normalized = String(thumbUri || "").trim();
  if (!normalized) {
    return "";
  }
  if (/^https?:\/\//i.test(normalized)) {
    return normalized;
  }
  return `${COVER_ORIGIN_URL}${normalized.replace(/^\/+/, "")}`;
};

const pickCoverUrl = (data) => {
  const candidates = [
    common.getString(data, "thumb_url", ""),
    common.getString(data, "cover_url", ""),
    buildOriginCoverUrl(common.getString(data, "thumb_uri", "")),
  ];

  for (const candidate of candidates) {
    if (isDisplayableCoverUrl(candidate)) {
      return candidate;
    }
  }

  return "";
};

const parseBook = (data) => ({
  book_id: common.getString(data, "book_id", ""),
  title: common.getString(data, "book_name", common.getString(data, "title", "")),
  author: common.getString(data, "author", ""),
  cover_url: pickCoverUrl(data),
  abstract: common.getString(data, "abstract", ""),
  category: common.getString(data, "category", ""),
  word_count: common.getString(data, "word_number", common.getString(data, "word_count", "")),
  score: common.getNumber(data, "score", 0),
  gender: common.getNumber(data, "gender", 0),
  creation_status: common.getNumber(data, "creation_status", 0),
  last_chapter_title: common.getString(data, "last_chapter_title", ""),
  last_update_time: common.getNumber(data, "last_chapter_update_time", 0),
});

const ensureApiKey = async () => {
  if (!ctx.api_key) {
    ctx.api_key = await common.requireEnv("FANQIE_APIKEY");
  }
  return ctx.api_key;
};

const buildUrl = async (typeId, params) => {
  const query = {
    apikey: await ensureApiKey(),
    type: typeId,
    ...(params || {}),
  };
  return common.appendQuery(API_URL, query);
};

const parseUserInfo = (html) => {
  const match = String(html || "").match(USER_RE);
  if (!match) {
    return {};
  }

  try {
    return JSON.parse(match[1]);
  } catch (_error) {
    return {};
  }
};

const getAuthMarkers = (html) => {
  const text = String(html || "");
  return {
    invalidKey: text.includes(INVALID_KEY_MARKER),
    logout: text.includes("?logout=1"),
    user: text.includes("const user ="),
    keyword: text.includes('name="keyword"'),
    download: text.includes("downloadBtn"),
  };
};

const isAuthenticatedHtml = (markers) =>
  markers.user || (markers.logout && (markers.keyword || markers.download));

const assertAuthenticatedHtml = (html) => {
  const markers = getAuthMarkers(html);
  const text = String(html || "");
  if (text.includes(INVALID_KEY_MARKER)) {
    ctx.logged_in = false;
    host.config_error("番茄书源登录失败：API Key 无效");
  }

  if (!isAuthenticatedHtml(markers)) {
    throw new Error("__novel_data_error__:fanqie login did not return authenticated html");
  }
};

const requireLoggedIn = () => {
  if (!ctx.logged_in) {
    host.config_error("当前番茄书源尚未登录；请先登录后再启用批量下载");
  }
};

module.exports = {
  manifest: {
    id: "fanqie",
    name: "番茄小说",
    version: "1.3.2",
    author: "novel-downloader-tui",
    description: "默认番茄小说书源插件，会自行从环境变量或 .env 中读取 FANQIE_APIKEY",
    required_envs: ["FANQIE_APIKEY"],
    optional_envs: [],
  },

  async configure() {
    ctx.api_key = await common.requireEnv("FANQIE_APIKEY");
    ctx.logged_in = false;
    ctx.user_info = null;
  },

  async login() {
    const apiKey = await ensureApiKey();
    const response = await common.postForm(
      LOGIN_URL,
      { apikey: apiKey },
      {},
      30,
      false,
    );

    if (response.status >= 400) {
      throw new Error(`login request failed with status ${response.status}`);
    }
    if (String(response.body || "").includes(INVALID_KEY_MARKER)) {
      ctx.logged_in = false;
      host.config_error("番茄书源登录失败：API Key 无效");
    }

    await host.log_info(`fanqie login post status=${response.status}`);
    const verifyResponse = await host.http_request({
      method: "GET",
      url: LOGIN_URL,
      timeout_seconds: 30,
    });
    await host.log_info(`fanqie login verify status=${verifyResponse.status}`);
    const html = String(verifyResponse.body || "");
    const markers = getAuthMarkers(html);
    await host.log_info(
      `fanqie login markers logout=${markers.logout} user=${markers.user} keyword=${markers.keyword} download=${markers.download} invalid=${markers.invalidKey}`,
    );
    assertAuthenticatedHtml(html);

    // Rain Web 当前会先下发 server_name_session，再在验证页分配 PHPSESSID。
    // 对 batch.php 来说，仅有验证页生成的 PHPSESSID 还不够；需要再用现有 cookie
    // 补一次 apikey POST，才能把下载能力绑定到当前 PHP session。
    const bindResponse = await common.postForm(
      LOGIN_URL,
      { apikey: apiKey },
      {},
      30,
      false,
    );
    await host.log_info(`fanqie login bind status=${bindResponse.status}`);
    if (bindResponse.status >= 400) {
      throw new Error(`login bind request failed with status ${bindResponse.status}`);
    }

    ctx.user_info = parseUserInfo(html);
    ctx.logged_in = true;
    return true;
  },

  async search(keywords, page) {
    const root = await common.getJson(await buildUrl(1, {
      keywords,
      page: page * 10,
    }));

    const results = [];
    if (Array.isArray(root.search_tabs)) {
      for (const tab of root.search_tabs) {
        if (!Array.isArray(tab?.data)) {
          continue;
        }
        for (const item of tab.data) {
          const book = item?.book_data?.[0];
          if (book) {
            results.push(parseBook(book));
          }
        }
        if (results.length > 0) {
          break;
        }
      }
      return results;
    }

    if (Array.isArray(root.data)) {
      for (const item of root.data) {
        const book = item?.book_data?.[0] ?? item;
        results.push(parseBook(book));
      }
    }

    return results;
  },

  async get_book_info(bookId) {
    const root = await common.getJson(await buildUrl(2, {
      bookid: bookId,
    }));
    if (!root.data) {
      return null;
    }
    return parseBook(root.data);
  },

  async get_toc(bookId) {
    const root = await common.getJson(await buildUrl(3, {
      bookid: bookId,
    }));

    const items = root?.data?.item_data_list;
    if (!Array.isArray(items)) {
      return [];
    }

    return items.map((item) => ({
      item_id: common.getString(item, "item_id", ""),
      title: common.getString(item, "title", ""),
      volume_name: common.getString(item, "volume_name", ""),
      word_count: common.getNumber(item, "chapter_word_number", 0),
      update_time: common.getNumber(item, "first_pass_time", 0),
    }));
  },

  async get_chapter(_bookId, itemId) {
    const root = await common.getJson(await buildUrl(4, {
      itemid: itemId,
    }));
    if (!root?.data?.content) {
      return null;
    }
    return {
      item_id: itemId,
      title: "",
      content: String(root.data.content),
    };
  },

  async get_batch_count(bookId) {
    requireLoggedIn();
    return rainBatch.getBatchCount(WEB_BASE_URL, bookId);
  },

  async get_batch(bookId, batchNo) {
    requireLoggedIn();
    return rainBatch.getBatch(WEB_BASE_URL, bookId, batchNo);
  },
};
